### DISCUSSION TITLE
Returning invalid/values which cannot be harmonized

### DISCUSSION THREAD

**Discussion by jakenoble** [CBIIT/ccdi-federation-api#5] [API Design Discussion]
We want to consider how should the API return values which cannot be harmonized to a CDE value, thus a `non-value` value?

The issue of being unable to map values to permissible CDE values will become more of an issue with `diagnosis`, for example.

In this situation, what are the group's thoughts on how the API should respond? Possible responses discussed so far;

- Return some form of `null`
- Return both the raw and harmonized value
- Return a custom error stating the raw value

**Comment by claymcleod** [comment]
In my recent [`/subject`](https://github.com/CBIIT/ccdi-federation-api/discussions/7) proposal, I opted to create a `harmonized` key under each of the metadata values:

```
{
  "id": 1,
  "name": "SubjectIdentifier001",
  "kind": "Individual",
  "metadata": {
    "sex": {
      "value": "Female",
      "harmonized": [
        {
          "standard": "CDE-6343385-v1",
          "value": "F"
        }
      ]
    },
    "race": [
      {
        "value": "American Indian or Alaska Native",
        "harmonized": [
          {
            "standard": "CDE-2192199-v1",
            "value": "American Indian or Alaska Native"
          }
        ]
      }
    ],
    "ethnicity": {
      "value": "Not Hispanic or Latino",
      "harmonized": [
        {
          "standard": "CDE-2192217-v2",
          "value": "Not Hispanic or Latino"
        }
      ]
    },
    "participant_id": [
      {
        "value": "SubjectIdentifier001"
      },
      {
        "value": "Other ID"
      }
    ]
  }
}
```

As you can see in my example, the participant identifiers do not have a harmonized value. Instead of including keys in the return object with a value of `null`, we can simply omit any entries from the harmonized array. The reason I like this approach is that we don't overload the meaning of `null`:

- `null` represents the data does not exist or was otherwise not collected.
- Not including entries in the `harmonized` list represents a term which cannot be harmonized to a particular controlled vocabulary.

**Comment by claymcleod** [comment]
This has been addressed in the latest iteration of the API, and thus, I'm going to close this discussion.
